pkg/services/provisioning/datasources/datasources.go

```diff
@@ -58,7 +58,7 @@ func newDatasourceProvisioner(log log.Logger, store Store, correlationsStore Cor
 }
 
 func (dc *DatasourceProvisioner) apply(ctx context.Context, cfg *configs) error {
-	if err := dc.deleteDatasources(ctx, cfg.DeleteDatasources); err != nil {
+	if err := dc.deleteDatasources(ctx, cfg.DeleteDatasources, cfg.Datasources); err != nil {
```

Comment:
this is pretty hard to follow imho, can we move the diff logic outside of `deleteDatasources` so that it only deletes without performing any extra logic that does not belong there?

we could probably create an intermediary slice that also contains `SkipPublish` and simply use that in`deleteDatasources`

Reply:
> so that it only deletes without performing any extra logic that does not belong there?

We still need to perform clean-up logic for data sources that were actually deleted. We cannot do it by removing correlations directly in the provisioning code, we need to actually publish DeleteDatasource event which triggers the clean-up at module level. This is needed not only to avoid code duplication but to support cases when more clean-up handlers are added in the code.

My first thought was that we can pass SkipPublish flag to the wrapper that deletes all data sources. When SkipPublish is set it would return a list of events that has to be published manually byt the consumer at the end. During provisioning we can remove from that list events for data sources that were not deleted, but re-created. I just thought it would actually harder to follow (with skip flag pushed farther down to the store), but I can try it out. What do you think?

Reply:
i see what you mean and it makes sense, i tried moving this logic out but then you actually need to query the DB to get the datasource UID, and we probably only want to do that in the delete function otherwise a lot more refactoring would be needed.

however i can se a case in which this doesn't work: when the datasources to be added do not contain an explicit UID, when doing `uidsToBeAdded = append(uidsToBeAdded, ds.UID)` we append a bunch of `""`, but the UID generation is actually stable and the datasource is effectively being recreated, but we treat it as only deleted

Reply:
Good catch!

I can use `safeUIDFromName` to figure out missing UIDs  like:

```go
var uidsToBeAdded []string
for _, ds := range dsToBeAdded {
    if ds.UID != "" {
	uidsToBeAdded = append(uidsToBeAdded, ds.UID)
    } else {
	uidsToBeAdded = append(uidsToBeAdded, safeUIDFromName(ds.Name))
    }
}
```

An alternative option is to change the provisioning logic so it first generates a list off commands to be executed. Then when it invokes `DeleteDataSourceCommand` it passes SkipPublish if there's an insert coming down the list. That would require changing deletion a bit so it generates UID like `createInsertCommand` but I'd need to check what are other implications of doing that.